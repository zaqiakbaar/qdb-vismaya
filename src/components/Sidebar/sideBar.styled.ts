import styled from "styled-components";
import { Layout, Menu } from 'antd';

export const SidebarContainer = styled(Layout.Sider)`
  height: 100vh;
  width: 250px;
  background: #fbfcfe !important;
  display: flex;
  flex-direction: column;
  box-shadow: -4px 0 10px rgba(0, 0, 0, 0.2);

  .ant-menu {
    flex:1;;
  }

  .menu-toggle {
    font-size: 20px;
    color: white;
    cursor: pointer;
    margin-bottom: 12px;
    justify-content: flex-end;
    display: flex;
    padding: 16px;
    background-color: dodgerblue
  }

  .ant-menu-item,
  .ant-menu-submenu-title {
    font-weight: 500;
    border-radius: 6px;
  }
  .ant-menu-item {
    color: lightgray;
    margin: 0;
  }
  .ant-menu-item-active {
    color: dodgerblue;
    margin: 0;
    width: 100%;
  }
  .ant-menu-item-selected{
    color: dodgerblue;
    width: 100%;
    border-radius: 0;
  }
`;

export const StyledHeading = styled.h3`
  padding-left: 15px;
`;

export const StyledMenu = styled(Menu)`
  padding: "10px 0";
`;

export const StyledMenuItems = styled(Menu.Item)`
  padding-left: 15px;
`
