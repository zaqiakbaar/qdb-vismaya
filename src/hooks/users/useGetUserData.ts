import React from "react";
import { fetchers } from "../../api/fetchers/fetchers"
import { IUserResponse } from "../../api/types/user/user";
import { config } from "../../config/global";
import { generateRandomNumber } from "../../utils/generateRandomNumber";

/**
 * 
 * Hooks to retrieve user details
 * 
 * @returns {IUserResponse | undefined} `userDetails` and `handleGetUserInformations` func
 */
export const useGetUserDetails = () => {

    const [userDetails, setUserDetails] = React.useState<IUserResponse | undefined>();

    const handleGetUserInformations = React.useCallback(async () => {
        const randomUserId = generateRandomNumber();
        try {
            const response = await fetchers<IUserResponse,unknown>({
                url: `${config.BASE_API_URL}/users/${randomUserId}`,
            });
            if(response?.id){
                setUserDetails(response);
            };
        } catch {
            //TODO - integrate liogs
        }
    },[]); 

    React.useEffect(() => {
        handleGetUserInformations();
    },[handleGetUserInformations])

    return { handleGetUserInformations, userDetails }
}