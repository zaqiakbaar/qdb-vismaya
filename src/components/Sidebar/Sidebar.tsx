import { Flex, Menu } from "antd";
import { DEFAULT_SELECTED_KEYS_NAVIGATION, menuItems } from "../../config/navigations/_menu";
import React from "react";
import { UserInfo } from "../UserInfo/UserInfo";
import { MenuOutlined } from "@ant-design/icons";
import { useNavigate } from "react-router-dom";
import { SidebarContainer, StyledHeading, StyledMenu, StyledMenuItems } from "./sideBar.styled";
import { useGetUserDetails } from "../../hooks/users/useGetUserData";

const { SubMenu } = Menu;

const Sidebar: React.FC = () => {
  const { userDetails: userInfo } = useGetUserDetails();
  const navigate = useNavigate();


  return (
    <SidebarContainer>
      <MenuOutlined className="menu-toggle" />
      {userInfo?.id && <UserInfo userDetails={userInfo} />}
      <StyledMenu mode="inline" defaultSelectedKeys={DEFAULT_SELECTED_KEYS_NAVIGATION} >
        {menuItems.map((item, index) => {
          return (
            <React.Fragment key={`menu-${index}`}>
              {item?.expandable ? (
                <SubMenu key={item.key} title={item.label} />
              ) : (<StyledHeading key={item.key}>{item.label}</StyledHeading>)}
              {(item.children || []).map((subItem) => (
                <StyledMenuItems key={subItem.key} onClick={() => navigate(subItem?.route)}>
                  <Flex align="center" gap={8}>
                    {subItem?.icon}
                    {subItem.label}
                  </Flex>
                </StyledMenuItems>
              )
              )}
            </React.Fragment>
          )
        }
        )}
      </StyledMenu>
    </SidebarContainer>
  );
};

export default Sidebar;

