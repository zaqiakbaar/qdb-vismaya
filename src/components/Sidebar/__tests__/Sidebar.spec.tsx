import { render, screen, waitFor } from "@testing-library/react";
import { useGetUserDetails } from "../../../hooks/users/useGetUserData";
import Sidebar from '../Sidebar';
import { BrowserRouter } from "react-router-dom";

jest.mock("../../../hooks/users/useGetUserData");

describe("it should render sidebar", () => {
  test("renders user info when userDetails is available", async () => {
    (useGetUserDetails as jest.Mock).mockReturnValue({
      userDetails: { id: 1, name: "Test User" },
    });

    render(
      <BrowserRouter>
        <Sidebar />
      </BrowserRouter>
    );
    await waitFor(() => expect(screen.getByText("Test User")));
    expect(screen.getByText("Test User")).toMatchSnapshot();
  });

  test("should not render user info when userDetails is unavailable", () => {
    (useGetUserDetails as jest.Mock).mockReturnValue({
      userDetails: undefined
    });

    const renderPages = render(
      <BrowserRouter>
        <Sidebar />
      </BrowserRouter>
    );


    expect(renderPages.queryByText("Test User")).toBeNull();
  });


});
