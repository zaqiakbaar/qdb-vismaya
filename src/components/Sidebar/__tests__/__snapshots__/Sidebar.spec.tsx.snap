// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should render sidebar renders user info when userDetails is available 1`] = `
<span
  class="ant-typography sc-fLDLck ggdxqX css-dev-only-do-not-override-xex5fb"
  textalign="center"
>
  Test User
</span>
`;
